scripts/format_warning_report.py

```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python3
+"""
+Take a '.report.json' output from pytest-json-report, and generate a
+markdown formatted report of any warnings generated.
+
+The script can also compare one JSON report to another, 'baseline',
+report and generate a markdown report that lists new warnings.
+
+In compare mode, the exit code is the number of new warnings. In normal
+mode, the exit code is the number of warnings. A negative exit code
+indicates an error.
+"""
+
+import argparse
+import json
+from dataclasses import dataclass
+from typing import Any, Dict, Iterable, List, Set, Tuple
+
+INDENT_SIZE = 2
+PROG = "format_warning_report"
+WARNINGS_KEY = "warnings"
+
+
+@dataclass(frozen=True, eq=True)
+class Warning:
+    """Stores information from a Python warning."""
+
+    message: str
+    category: str
+    when: str
+    filename: str
+    lineno: int
+
+    def __str__(self) -> str:
+        """Convert to formatted string."""
+        return f"{self.filename}:{self.lineno}: {self.category}: {self.message}"
+
+
+def parse_args(sys_args: List[str]) -> argparse.Namespace:
+    """Parse command line options."""
+    parser = argparse.ArgumentParser(PROG, description=__doc__)
+    parser.add_argument("report_file", help="path to .report.json to parse")
+    parser.add_argument(
+        "--compare",
+        default=None,
+        help="path to .report.json to compare warnings to",
+        metavar="baseline_report_file",
+    )
+    return parser.parse_args(sys_args)
+
+
+def load_warnings(file_path: str) -> Set[Warning]:
+    """Load a set of warnings from file."""
+    with open(file_path, "r") as f:
+        data = json.load(f)
+    try:
+        return set([Warning(**warning) for warning in data[WARNINGS_KEY]])
+    except KeyError:  # no warnings found
+        return set()
+
+
+def format_warnings_list(warnings: Iterable[Warning]) -> List[str]:
+    """Format the list of warnings into lines of markdown."""
+    whens: Dict[str, List[Warning]] = {}
+    for warning in warnings:
+        try:
+            whens[warning.when].append(warning)
+        except KeyError:
+            whens[warning.when] = [warning]
+    lines = []
+    for when, warns in whens.items():
+        lines.append(f"#### On {when}\n")
+        for warn in warns:
+            lines.append(f"- `{warn}`")
+        lines.append("")
+    return lines[:-1]  # we do not need a trailing new line
```

Comment:
This is probably a bit more robust? Don't want to remove too much
```suggestion
    return lines.strip("\n")  # we do not need a trailing new line
```

Reply:
I'm building up the lines as a list, so we can't use `strip` here. I think the slicing is robust in this case, as the last line to be appended in the loop is always empty.